File: 077__Combinations.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """

        def helper(tlist, start, end):
            if len(tlist) == k:
                result.append(tlist)
                return
            for i in range(start, end + 1):
                tlist.append(i)
                helper(tlist, i + 1, end)
                tlist.pop()

        result = []
        helper([], 1, n)
        return result

File: Java/077__Combinations.java
class Solution {

    List<List<Integer>> result = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        helper(n, 1, k, new ArrayList<>());
        return result;
    }

    private void helper(int n, int index, int k, List<Integer> cur_arr) {
        if (k == 0) result.add(new ArrayList<>(cur_arr));
        if (index > n) return;
        for (int x = index; x <= n; x++) {
            cur_arr.add(x);
            helper(n, x + 1, k - 1, cur_arr);
            cur_arr.remove(cur_arr.size() - 1);
        }
    }
}
